File: HTML5/game.js
var game = new Game();

function Game()
{
   this.whiteTurn = true;
   this.possibleMoves = [];
   this.possiblePos = [];

   this.ismarked;
   this.marked = new Position(0,0);

   this.checkmate;
   this.check;
   this.gamedraw;
   
   this.reset = function()
   {
      board.reset();
      this.ismarked = false;
      this.whiteTurn = true;
      this.check = false;
      this.checkmate = false;
      this.gamedraw = false;
      this.possibleMoves = rules.checkMoves(board, this.whiteTurn, true);
      if (!settings.playerWhite)
      {
         opponent.move(board, this.possibleMoves, this.whiteTurn, settings.difficulty);
         this.whiteTurn = !this.whiteTurn;
         this.possibleMoves = rules.checkMoves(board, this.whiteTurn, true);
      }
   }
   
   this.boardclick = function(call_info, bpos)
   {
      var height = canvas.height - 50;
      var width = canvas.width;
      if (height > width) height = width;
      if (width > height) width = height;   
      var rectsize = Math.floor(height / 8);
      
      var posx = Math.floor(bpos[0] / rectsize);
      var posy = Math.floor(bpos[1] / rectsize);
      if (!settings.playerWhite) 
      {
         posx = 7 - posx;
         posy = 7 - posy;
      }
      this.tryMove(posx, posy);
      
      if (this.gamedraw) maingui.tabs[1].name = "Game: Draw";
      else if (this.checkmate) maingui.tabs[1].name = "Game: Checkmate";
      else if (this.check) maingui.tabs[1].name = "Game: Check";
      else maingui.tabs[1].name = "Game";
   }
   this.buttonBoard = new Button(this.boardclick, this, 0);
   
   this.tryMove = function(x, y)
   {
      if (this.checkmate) return;
            
      if (this.posIsPlayer(x, y)) this.ismarked = false;
      
      if (this.ismarked)
      {
         for (var i = 0; i < this.possiblePos.length; i++)
         {
            if (this.possiblePos[i].x == x && this.possiblePos[i].y == y)
            {
               board.move(new Move(this.marked, new Position(y, x)), true);
               this.whiteTurn = !this.whiteTurn;
               this.possibleMoves = rules.checkMoves(board, this.whiteTurn, true);
               if (settings.playerWhite != this.whiteTurn && this.possibleMoves.length > 0) 
               { 
                  opponent.move(board, this.possibleMoves, this.whiteTurn, settings.difficulty);
                  this.whiteTurn = !this.whiteTurn;
                  this.possibleMoves = rules.checkMoves(board, this.whiteTurn, true);
               }
               this.checkCheck();
               break;
            }
         }
         this.ismarked = false;
      }
      else
      {
         var start = 7;
         if (settings.playerWhite) start = 1;
         if (board.pieces[y][x] >= start && board.pieces[y][x] < start + 6)
         {
            this.possiblePos = []
            for (var i = 0; i < this.possibleMoves.length; i++)
            {
               if (this.possibleMoves[i].source.x == x && this.possibleMoves[i].source.y == y)
               {
                  this.possiblePos.push(this.possibleMoves[i].target);
               }
            }
            if (this.possiblePos.length > 0)
            {
               this.ismarked = true;
               this.marked.x = x;
               this.marked.y = y;
            }
         }
      }
   }
   
   this.posIsPlayer = function(x, y)
   {
      if (this.whiteTurn && board.pieces[y][x] >= 1 && board.pieces[y][x] < 7) return true;
      if (!this.whiteTurn && board.pieces[y][x] >= 7 && board.pieces[y][x] < 13) return true;
      return false;
   }

   this.draw = function(x, y, sizeX, sizeY)
   {  
      var height = canvas.height - 50;
      var width = canvas.width;
      if (height > width) height = width;
      if (width > height) width = height;   
      this.buttonBoard.draw(0, 50, width, height, "", "red", "salmon", "black");
      
      drawer.draw(x, y, sizeX, sizeY)
   }
   
   this.checkCheck = function()
   {
      this.check = !rules.checkFieldSave(board, this.whiteTurn, rules.getKingPos(board, this.whiteTurn));
      this.checkmate = this.possibleMoves.length == 0 && this.check;
      this.gamedraw = this.possibleMoves.length == 0 && !this.check;
   }
  
} 

File: Csharp/ChesszumSpass/Game.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ChesszumSpass
{
    class Game
    {
        public Board board;

        int difficulty;
        bool playerWhite;
        bool playerBoth;
        public bool whiteTurn;

        public List<Move> possibleMoves;
        public List<Position> possiblePos;

        public bool ismarked;
        public Position marked;

        public bool checkmate;
        public bool check;
        public bool draw;

        public Game()
        {
            board = new Board();
            marked = new Position(-1, -1);
            possiblePos = new List<Position>();
            possibleMoves = new List<Move>();
            playerWhite = true;
            reset();
        }

        public void reset()
        {
            board.reset();
            ismarked = false;
            whiteTurn = true;
            check = false;
            checkmate = false;
            draw = false;
            if (playerBoth) playerWhite = true;
            possibleMoves = Rules.checkMoves(board, whiteTurn, true);
            if (playerWhite != whiteTurn)
            {
                Opponent.move(board, possibleMoves, whiteTurn, difficulty);
                whiteTurn = !whiteTurn;
                possibleMoves = Rules.checkMoves(board, whiteTurn, true);
            }
        }

        public void setPlayer(bool isWhite, bool both)
        {
            playerWhite = isWhite;
            playerBoth = both;
        }

        public void setDifficulty(int difficulty)
        {
            this.difficulty = difficulty;
        }

        public void tryMove(int x, int y)
        {
            if (checkmate) return;
            
            if (posIsPlayer(x, y)) ismarked = false;

            if (ismarked)
            {
                for (int i = 0; i < possiblePos.Count; i++)
                {
                    if (possiblePos[i].x == x && possiblePos[i].y == y)
                    {
                        board.move(new Move(marked, new Position(y, x)), true);
                        whiteTurn = !whiteTurn;
                        if (playerBoth) playerWhite = !playerWhite;
                        possibleMoves = Rules.checkMoves(board, whiteTurn, true);
                        if (playerWhite != whiteTurn && possibleMoves.Count > 0) 
                        { 
                            Opponent.move(board, possibleMoves, whiteTurn, difficulty);
                            whiteTurn = !whiteTurn;
                            possibleMoves = Rules.checkMoves(board, whiteTurn, true);
                        }
                        checkCheck();
                        break;
                    }
                }
                ismarked = false;
            }
            else
            {
                int start = 7;
                if (playerWhite) start = 1;
                if ((int)board.pieces[y, x] >= start && (int)board.pieces[y, x] < start + 6)
                {
                    possiblePos.Clear();
                    for (int i = 0; i < possibleMoves.Count; i++)
                    {
                        if (possibleMoves[i].source.x == x && possibleMoves[i].source.y == y)
                        {
                            possiblePos.Add(possibleMoves[i].target);
                        }
                    }
                    if (possiblePos.Count > 0)
                    {
                        ismarked = true;
                        marked.x = x;
                        marked.y = y;
                    }
                }
            }
        }

        public void revertStep()
        {
            if (board.revertStep(true))
            {
                ismarked = false;
                whiteTurn = !whiteTurn;
                if (playerBoth) playerWhite = !playerWhite;
                possibleMoves = Rules.checkMoves(board, whiteTurn, true);
                checkCheck();
            }
        }

        private bool posIsPlayer(int x, int y)
        {
            if (playerWhite && (int)board.pieces[y, x] >= 1 && (int)board.pieces[y, x] < 7) return true;
            if (!playerWhite && (int)board.pieces[y, x] >= 7 && (int)board.pieces[y, x] < 13) return true;
            return false;
        }

        private void checkCheck()
        {
            check = !Rules.checkFieldSave(board, whiteTurn, Rules.getKingPos(board, whiteTurn));
            checkmate = possibleMoves.Count == 0 && check;
            draw = possibleMoves.Count == 0 && !check;
        }

    }
}
